scene/main/viewport.cpp

```diff
@@ -1351,7 +1351,12 @@ Ref<InputEvent> Viewport::_make_input_local(const Ref<InputEvent> &ev) {
 
 Vector2 Viewport::get_mouse_position() const {
 	ERR_READ_THREAD_GUARD_V(Vector2());
-	if (DisplayServer::get_singleton()->has_feature(DisplayServer::FEATURE_MOUSE)) {
+	if (!is_inside_tree() || (Object::cast_to<SubViewport>(this) && !Object::cast_to<SubViewportContainer>(get_parent()))) {
```

Comment:
I'm wondering if this condition can be rewritten based on some already existing is-attached-to-screen functionality or at least should be put in a new `virtual bool is_attached_to_screen()`, or the like.

Maybe not. It's that checking the type of `this` seems a bit odd.

Reply:
I am not aware of any existing functionality, that covers this.

`is_attached_to_screen` could be something like:
```c++
bool SubViewport::is_attached_to_screen() const {
   return Object::cast_to<SubViewportContainer>(get_parent()) && get_parent()->get_viewport() && get_parent()->get_viewport()->is_attached_to_screen();
}
bool Window::is_attached_to_screen() const {
    if (get_embedder()) {
        return get_embedder()->is_attached_to_screen();
    }
    return is_inside_tree();
}
```

Alternatively we could introduce a virtual function:
``` c++
bool Viewport::is_window() const {
    return true;
}
bool SubViewport::is_window() const {
    return false;
}
```
And replace `Object::cast_to<SubViewport>(this)` by `!_is_window()`.

Reply:
I have updated the PR and introduced a `is_directly_attached_to_screen` method.